koschei/backend/repo_cache.py

```diff
@@ -19,157 +19,44 @@
-        are not parseable as repo descriptor strings. """
-        log.debug('Reading cached repos from disk...')
-        for repo_descriptor, repo_path in [(RepoDescriptor.from_string(repo_name),
-                                            os.path.join(self._repos_dir, repo_name))
-                                           for repo_name in os.listdir(self._repos_dir)]:
-            if repo_descriptor:
-                yield repo_descriptor, repo_path
-            elif os.path.exists(repo_path):
-                # Remove directories that don't look like valid repos,
-                # such as incomplete downloads.
-                shutil.rmtree(repo_path)
-                log.debug('Removed bogus file from cache: {}'.format(repo_path))
-
-    def _ensure_capacity(self):
-        """ Make sure there is enough room for new repo to be added. """
-        if len(self._repos) == self._capacity:
-            # Adding repo would exceed capacity, discard LRU repo to free space.
-            _, repo_path = self._repos.popitem(last=False)
-            if os.path.exists(repo_path):
-                shutil.rmtree(repo_path)
-        assert len(self._repos) < self._capacity
-
-    def _download_repo(self, repo_descriptor, repo_dir):
-        """ Download repo from Koji to disk. """
-        h = librepo.Handle()
-        h.destdir = repo_dir
-        # pylint:disable=no-member
-        h.repotype = librepo.LR_YUMREPO
-        h.urls = [repo_descriptor.make_url()]
-        h.yumdlist = ['primary', 'filelists', 'group', 'group_gz']
-        h.perform(librepo.Result())
-
-    def _load_sack(self, repo_descriptor, repo_path, build_cache=False):
-        """ Load repo from disk into memory as sack. """
-        cache_dir = os.path.join(repo_path, 'cache')
-        for_arch = get_config('dependency.resolve_for_arch')
-        if build_cache:
-            os.mkdir(cache_dir)
-        sack = hawkey.Sack(arch=for_arch, cachedir=cache_dir)
-        h = librepo.Handle()
-        h.local = True
-        # pylint:disable=no-member
-        h.repotype = librepo.LR_YUMREPO
-        h.urls = [repo_path]
-        h.yumdlist = ['primary', 'filelists', 'group']
-        repodata = h.perform(librepo.Result()).yum_repo
-        repo = hawkey.Repo(str(repo_descriptor))
-        repo.repomd_fn = repodata['repomd']
-        repo.primary_fn = repodata['primary']
-        repo.filelists_fn = repodata['filelists']
-        sack.load_yum_repo(repo, load_filelists=True, build_cache=build_cache)
+        self.log = logging.getLogger('koschei.repo_cache.RepoCache')
+        super(RepoCache, self).__init__(
+            cachedir=os.path.join(get_config('directories.cachedir'), 'repodata'),
+            capacity=get_config('dependency.cache_l2_capacity'),
+            log=self.log,
+        )
+
+    # @Override
+    def read_item(self, repo_descriptor, cachedir):
+        return repo_util.load_sack(cachedir, repo_descriptor)
```

Comment:
That's a very good point. I shouldn't make assumptions about libsolv's internal file handling (while it would probably work this way, it may keep the descriptor open, which may in turn violate the cache size constraint). I'll change it to context manager that keeps the read lock while the sack is being used.